consensus/istanbul/backend/backend.go

```diff
@@ -794,3 +769,48 @@ func (sb *Backend) ConnectToVals() {
 		sb.RefreshValPeers(valset)
 	}
 }
+
+func (sb *Backend) retrieveRegisteredAndElectedValidators() (map[common.Address]bool, error) {
+	sb.cachedRegisteredElectedValSetMu.Lock()
+	defer sb.cachedRegisteredElectedValSetMu.Unlock()
+
+	// Check to see if there is a cached registered/elected validator set, and if it's for the current block
+	if sb.cachedRegisteredElectedValSet != nil && time.Since(sb.cachedRegisteredElectedValSetTimestamp) <= 1*time.Minute {
+		return sb.cachedRegisteredElectedValSet, nil
```

Comment:
Returning this map isn't quite thread safe is it? If we want to keep things as safe as possible we should make a copy of this before returning